components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaAccountManager.kt

```diff
@@ -0,0 +1,423 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.fxa
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.async
+import java.util.concurrent.ConcurrentLinkedQueue
+import kotlin.coroutines.CoroutineContext
+
+import mozilla.components.support.base.log.Log
+import mozilla.components.support.base.observer.Observable
+import mozilla.components.support.base.observer.ObserverRegistry
+import java.io.Closeable
+import java.util.concurrent.Executors
+
+enum class AccountState {
+    Start,
+    NotAuthenticated,
+    AuthenticatedNoProfile,
+    AuthenticatedWithProfile
+}
+
+/**
+ * Base class for account events.
+ * Events aren't a simple enum class because we might want to pass data along with some of the events.
+ */
+sealed class AccountEvent {
```

Comment:
Nit: A state-machine event can be mistaken for a more classic observable-pattern event. It took me a little bit of time to realize these events were just consumed by the state machine. Could we use another term such as `Input`?